File: VocaDbModel/DataContracts/ReleaseEvents/ReleaseEventContract.cs
#nullable disable

using System;
using VocaDb.Model.DataContracts.Songs;
using VocaDb.Model.DataContracts.Venues;
using VocaDb.Model.Domain;
using VocaDb.Model.Domain.Globalization;
using VocaDb.Model.Domain.Images;
using VocaDb.Model.Domain.ReleaseEvents;
using VocaDb.Model.Helpers;

namespace VocaDb.Model.DataContracts.ReleaseEvents
{
	public class ReleaseEventContract : IReleaseEvent, IEntryImageInformation, IEntryBase, IEntryWithStatus
	{
		EntryType IEntryImageInformation.EntryType => EntryType.ReleaseEvent;
		EntryType IEntryBase.EntryType => EntryType.ReleaseEvent;
		string IEntryImageInformation.Mime => PictureMime;
		ImagePurpose IEntryImageInformation.Purpose => ImagePurpose.Main;
		string IEntryBase.DefaultName => Name;

		public ReleaseEventContract()
		{
			Description = string.Empty;
		}

#nullable enable
		public ReleaseEventContract(ReleaseEvent ev, ContentLanguagePreference languagePreference, bool includeSeries = false, bool includeSeriesLinks = false)
			: this()
		{
			ParamIs.NotNull(() => ev);

			AdditionalNames = ev.Names.GetAdditionalNamesStringForLanguage(languagePreference);
			Category = ev.Category;
			CustomName = ev.CustomName;
			Date = ev.Date;
			Deleted = ev.Deleted;
			Description = ev.Description;
			EndDate = ev.EndDate;
			Id = ev.Id;
			Name = ev.TranslatedName[languagePreference];
			PictureMime = ev.PictureMime;
			SongList = ObjectHelper.Convert(ev.SongList, s => new SongListBaseContract(s));
			Status = ev.Status;
			UrlSlug = ev.UrlSlug;
			Venue = ObjectHelper.Convert(ev.Venue, v => new VenueContract(v, languagePreference, includeSeriesLinks));
			VenueName = ev.VenueName;
			Version = ev.Version;

			if (includeSeries && ev.HasSeries)
			{
				Series = new ReleaseEventSeriesContract(ev.Series, languagePreference, includeSeriesLinks);
			}
		}
#nullable disable

		public string AdditionalNames { get; init; }

		public EventCategory Category { get; init; }

		public bool CustomName { get; set; }

		public DateTime? Date { get; init; }

		public bool Deleted { get; init; }

		public string Description { get; init; }

		public DateTime? EndDate { get; init; }

		public bool HasVenueOrVenueName => Venue != null || !string.IsNullOrEmpty(VenueName);

		public int Id { get; set; }

		public EventCategory InheritedCategory => Series?.Category ?? Category;

		public string Name { get; init; }

		public string PictureMime { get; init; }

		public ReleaseEventSeriesContract Series { get; set; }

		public SongListBaseContract SongList { get; init; }

		public EntryStatus Status { get; init; }

		public string UrlSlug { get; init; }

		public VenueContract Venue { get; init; }

		public string VenueName { get; init; }

		public int Version { get; init; }
	}
}

File: VocaDbWeb/Scripts/DataContracts/ReleaseEvents/ReleaseEventContract.ts
import { EntryThumbContract } from '@/DataContracts/EntryThumbContract';
import { LocalizedStringWithIdContract } from '@/DataContracts/Globalization/LocalizedStringWithIdContract';
import { PVContract } from '@/DataContracts/PVs/PVContract';
import { ArtistForEventContract } from '@/DataContracts/ReleaseEvents/ArtistForEventContract';
import { EventSeriesContract } from '@/DataContracts/ReleaseEvents/EventSeriesContract';
import { SongListBaseContract } from '@/DataContracts/SongListBaseContract';
import { TagUsageForApiContract } from '@/DataContracts/Tag/TagUsageForApiContract';
import { VenueForApiContract } from '@/DataContracts/Venue/VenueForApiContract';
import { WebLinkContract } from '@/DataContracts/WebLinkContract';
import { EntryStatus } from '@/Models/EntryStatus';
import { EventCategory } from '@/Models/Events/EventCategory';

// Matches ReleaseEventForApiContract
export interface ReleaseEventContract {
	additionalNames?: string;
	artists: ArtistForEventContract[];
	category: EventCategory;
	/**  DateOnly */
	date?: string;
	defaultNameLanguage: string;
	description?: string;
	/**  DateOnly */
	endDate?: string;
	id: number;
	mainPicture?: EntryThumbContract;
	name: string;
	names?: LocalizedStringWithIdContract[];
	pvs?: PVContract[];
	series?: EventSeriesContract;
	songList?: SongListBaseContract;
	status: EntryStatus;
	tags?: TagUsageForApiContract[];
	urlSlug?: string;
	venue?: VenueForApiContract;
	version?: number;
	venueName?: string;
	webLinks: WebLinkContract[];
}
